share/man/man1/pegasus-mpi-cluster.1
'\" t
.\"     Title: pegasus-mpi-cluster
.\"    Author: [see the "Author" section]
.\" Generator: DocBook XSL Stylesheets v1.76.1 <http://docbook.sf.net/>
.\"      Date: 11/25/2013
.\"    Manual: \ \&
.\"    Source: \ \&
.\"  Language: English
.\"
.TH "PEGASUS\-MPI\-CLUSTE" "1" "11/25/2013" "\ \&" "\ \&"
.\" -----------------------------------------------------------------
.\" * Define some portability stuff
.\" -----------------------------------------------------------------
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" http://bugs.debian.org/507673
.\" http://lists.gnu.org/archive/html/groff/2009-02/msg00013.html
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"
pegasus-mpi-cluster \- a tool for running computational workflows expressed as DAGs (Directed Acyclic Graphs) on computational clusters using MPI\&.
.SH "SYNOPSIS"
.sp
.nf
\fBpegasus\-mpi\-cluster\fR [options] \fIworkflow\&.dag\fR
.fi
.SH "DESCRIPTION"
.sp
\fBpegasus\-mpi\-cluster\fR is a tool used to run HTC (High Throughput Computing) scientific workflows on systems designed for HPC (High Performance Computing)\&. Many HPC systems have custom architectures that are optimized for tightly\-coupled, parallel applications\&. These systems commonly have exotic, low\-latency networks that are designed for passing short messages very quickly between compute nodes\&. Many of these networks are so highly optimized that the compute nodes do not even support a TCP/IP stack\&. This makes it impossible to run HTC applications using software that was designed for commodity clusters, such as Condor\&.
.sp
\fBpegasus\-mpi\-cluster\fR was developed to enable loosely\-coupled HTC applications such as scientific workflows to take advantage of HPC systems\&. In order to get around the network issues outlined above, \fBpegasus\-mpi\-cluster\fR uses MPI (Message Passing Interface), a commonly used API for writing SPMD (Single Process, Multiple Data) parallel applications\&. Most HPC systems have an MPI implementation that works on whatever exotic network architecture the system uses\&.
.sp
An \fBpegasus\-mpi\-cluster\fR job consists of a single master process (this process is rank 0 in MPI parlance) and several worker processes\&. The master process manages the workflow and assigns workflow tasks to workers for execution\&. The workers execute the tasks and return the results to the master\&. Any output written to stdout or stderr by the tasks is captured (see \fBTASK STDIO\fR)\&.
.sp
\fBpegasus\-mpi\-cluster\fR applications are expressed as DAGs (Directed Acyclic Graphs) (see \fBDAG FILES\fR)\&. Each node in the graph represents a task, and the edges represent dependencies between the tasks that constrain the order in which the tasks are executed\&. Each task is a program and a set of parameters that need to be run (i\&.e\&. a command and some optional arguments)\&. The dependencies typically represent data flow dependencies in the application, where the output files produced by one task are needed as inputs for another\&.
.sp
If an error occurs while executing a DAG that causes the workflow to stop, it can be restarted using a rescue file, which records the progress of the workflow (see \fBRESCUE FILES\fR)\&. This enables \fBpegasus\-mpi\-cluster\fR to pick up running the workflow where it stopped\&.
.sp
\fBpegasus\-mpi\-cluster\fR was designed to work either as a standalone tool or as a complement to the Pegasus Workflow Managment System (WMS)\&. For more information about using PMC with Pegasus see the section on \fBPMC AND PEGASUS\fR\&.
.sp
\fBpegasus\-mpi\-cluster\fR allows applications expressed as a DAG to be executed in parallel on a large number of compute nodes\&. It is designed to be simple, lightweight and robust\&.
.SH "OPTIONS"
.PP
\fB\-h\fR, \fB\-\-help\fR
.RS 4
Print help message
.RE
.PP
\fB\-V\fR, \fB\-\-version\fR
.RS 4
Print version information
.RE
.PP
\fB\-v\fR, \fB\-\-verbose\fR
.RS 4
Increase logging verbosity\&. Adding multiple
\fB\-v\fR
increases the level more\&. The default log level is
\fIINFO\fR\&. (see
\fBLOGGING\fR)
.RE
.PP
\fB\-q\fR, \fB\-\-quiet\fR
.RS 4
Decrease logging verbosity\&. Adding multiple
\fB\-q\fR
decreases the level more\&. The default log level is
\fIINFO\fR\&. (see
\fBLOGGING\fR)
.RE
.PP
\fB\-s\fR, \fB\-\-skip\-rescue\fR
.RS 4
Ignore the rescue file for
\fIworkflow\&.dag\fR
if it exists\&. Note that
\fBpegasus\-mpi\-cluster\fR
will still create a new rescue file for the current run\&. The default behavior is to use the rescue file if one is found\&. (see
\fBRESCUE FILES\fR)
.RE
.PP
\fB\-o\fR \fIpath\fR, \fB\-\-stdout\fR \fIpath\fR
.RS 4
Path to file for task stdout\&. (see
\fBTASK STDIO\fR
and
\fB\-\-per\-task\-stdio\fR)
.RE
.PP
\fB\-e\fR \fIpath\fR, \fB\-\-stderr\fR \fIpath\fR
.RS 4
Path to file for task stderr\&. (see
\fBTASK STDIO\fR
and
\fB\-\-per\-task\-stdio\fR)
.RE
.PP
\fB\-m\fR \fIM\fR, \fB\-\-max\-failures\fR \fIM\fR
.RS 4
Stop submitting new tasks after
\fIM\fR
tasks have failed\&. Once
\fIM\fR
has been reached,
\fBpegasus\-mpi\-cluster\fR
will finish running any tasks that have been started, but will not start any more tasks\&. This option is used to prevent
\fBpegasus\-mpi\-cluster\fR
from continuing to run a workflow that is suffering from a systematic error, such as a missing binary or an invalid path\&. The default for
\fIM\fR
is 0, which means unlimited failures are allowed\&.
.RE
.PP
\fB\-t\fR \fIT\fR, \fB\-\-tries\fR \fIT\fR
.RS 4
Attempt to run each task
\fIT\fR
times before marking the task as failed\&. Note that the
\fIT\fR
tries do not count as failures for the purposes of the
\fB\-m\fR
option\&. A task is only considered failed if it is tried
\fIT\fR
times and all
\fIT\fR
attempts result in a non\-zero exitcode\&. The value of
\fIT\fR
must be at least 1\&. The default is 1\&.
.RE
.PP
\fB\-n\fR, \fB\-\-nolock\fR
.RS 4
Do not lock DAGFILE\&. By default,
\fBpegasus\-mpi\-cluster\fR
will attempt to acquire an exclusive lock on DAGFILE to prevent multiple MPI jobs from running the same DAG at the same time\&. If this option is specified, then the lock will not be acquired\&.
.RE
.PP
\fB\-r\fR, \fB\-\-rescue\fR \fIpath\fR
.RS 4
Path to rescue log\&. If the file exists, and
\fB\-s\fR
is not specified, then the log will be used to recover the state of the workflow\&. The file is truncated after it is read and a new rescue log is created in its place\&. The default is to append
\fI\&.rescue\fR
to the DAG file name\&. (see
\fBRESCUE FILES\fR)
.RE
.PP
\fB\-\-host\-script\fR \fIpath\fR
.RS 4
Path to a script or executable to launch on each unique host that
\fBpegasus\-mpi\-cluster\fR
is running on\&. This path can also be set using the PMC_HOST_SCRIPT environment variable\&. (see
\fBHOST SCRIPTS\fR)
.RE
.PP
\fB\-\-host\-memory\fR \fIsize\fR
.RS 4
Amount of memory available on each host in MB\&. The default is to determine the amount of physical RAM automatically\&. This value can also be set using the PMC_HOST_MEMORY environment variable\&. (see
\fBRESOURCE\-BASED SCHEDULING\fR)
.RE
.PP
\fB\-\-host\-cpus\fR \fIcpus\fR
.RS 4
Number of CPUs available on each host\&. The default is to determine the number of CPU cores automatically\&. This value can also be set using the PMC_HOST_CPUS environment variable\&. (see
\fBRESOURCE\-BASED SCHEDULING\fR)
.RE
.PP
\fB\-\-strict\-limits\fR
.RS 4
This enables strict memory usage limits for tasks\&. When this option is specified, and a task tries to allocate more memory than was requested in the DAG, the memory allocation operation will fail\&.
.RE
.PP
\fB\-\-max\-wall\-time\fR \fIminutes\fR
.RS 4
This is the maximum number of minutes that
\fBpegasus\-mpi\-cluster\fR
will allow the workflow to run\&. When this time expires
\fBpegasus\-mpi\-cluster\fR
will abort the workflow and merge all of the stdout/stderr files of the workers\&. The value is in minutes, and the default is unlimited wall time\&. This option was added so that the output of a workflow will be recorded even if the workflow exceeds the max wall time of its batch job\&. This value can also be set using the PMC_MAX_WALL_TIME environment variable\&.
.RE
.PP
\fB\-\-per\-task\-stdio\fR
.RS 4
This causes PMC to generate a \&.out\&.XXX and a \&.err\&.XXX file for each task instead of writing task stdout/stderr to
\fB\-\-stdout\fR
and
\fB\-\-stderr\fR\&. The name of the files are "TASKNAME\&.out\&.XXX" and "TASKNAME\&.err\&.XXX", where "TASKNAME" is the name of the task from the DAG and "XXX" is a sequence number that is incremented each time the task is tried\&. This option overrides the values for
\fB\-\-stdout\fR
and
\fB\-\-stderr\fR\&. This argument is used by Pegasus when workflows are planned in PMC\-only mode to facilitate debugging and monitoring\&.
.RE
.PP
\fB\-\-jobstate\-log\fR
.RS 4
This option causes PMC to generate a jobstate\&.log file for the workflow\&. The file is named "jobstate\&.log" and is placed in the same directory where the DAG file is located\&. If the file already exists, then PMC appends new lines to the existing file\&. This option is used by Pegasus when workflows are planned in PMC\-only mode to facilitate monitoring\&.
.RE
.PP
\fB\-\-monitord\-hack\fR
.RS 4
This option causes PMC to generate a \&.dagman\&.out file for the workflow\&. This file mimics the contents of the \&.dagman\&.out file generated by Condor DAGMan\&. The point of this option is to trick monitord into thinking that it is dealing with DAGMan so that it will generate the appropriate events to populate the STAMPEDE database for monitoring purposes\&. The file is named "DAG\&.dagman\&.out" where "DAG" is the path to the PMC DAG file\&.
.RE
.PP
\fB\-\-no\-resource\-log\fR
.RS 4
Do not generate a
\fIworkflow\&.dag\&.resource\fR
file for the workflow\&.
.RE
.PP
\fB\-\-no\-sleep\-on\-recv\fR
.RS 4
Do not use polling with sleep() to implement message receive\&. (see
\fBKnown Issues: CPU Usage\fR)
.RE
.PP
\fB\-\-maxfds\fR
.RS 4
Set the maximum number of file descriptors that can be left open by the master for I/O forwarding\&. By default this value is set automatically based on the value of getrlimit(RLIMIT_NOFILE)\&. The value must be at least 1, and cannot be more than RLIMIT_NOFILE\&.
.RE
.SH "DAG FILES"
.sp
\fBpegasus\-mpi\-cluster\fR workflows are expressed using a simple text\-based format similar to that used by Condor DAGMan\&. There are only two record types allowed in a DAG file: \fBTASK\fR and \fBEDGE\fR\&. Any blank lines in the DAG (lines with all whitespace characters) are ignored, as are any lines beginning with # (note that # can only appear at the beginning of a line, not in the middle)\&.
.sp
The format of a \fBTASK\fR record is:
.sp
.if n \{\
.RS 4
.\}
.nf
"TASK" id [options\&.\&.\&.] executable [arguments\&.\&.\&.]
.fi
.if n \{\
.RE
.\}
.sp
Where \fIid\fR is the ID of the task, \fIoptions\fR is a list of task options, \fIexecutable\fR is the path to the executable or script to run, and \fIarguments\&...\fR is a space\-separated list of arguments to pass to the task\&. An example is:
.sp
.if n \{\
.RS 4
.\}
.nf
TASK t01 \-m 10 \-c 2 /bin/program \-a \-b
.fi
.if n \{\
.RE
.\}
.sp
This example specifies a task \fIt01\fR that requires 10 MB memory and 2 CPUs to run \fI/bin/program\fR with the arguments \fI\-a\fR and \fI\-b\fR\&. The available task options are:
.PP
\fB\-m\fR \fIM\fR, \fB\-\-request\-memory\fR \fIM\fR
.RS 4
The amount of memory required by the task in MB\&. The default is 0, which means memory is not considered for this task\&. This option can be set for a job in the DAX by specifying the pegasus::pmc_request_memory profile\&. (see
\fBRESOURCE\-BASED SCHEDULING\fR)
.RE
.PP
\fB\-c\fR \fIN\fR, \fB\-\-request\-cpus\fR \fIN\fR
.RS 4
The number of CPUs required by the task\&. The default is 1, which implies that the number of slots on a host should be less than or equal to the number of physical CPUs in order for all the slots to be used\&. This option can be set for a job in the DAX by specifying the pegasus::pmc_request_cpus profile\&. (see
\fBRESOURCE\-BASED SCHEDULING\fR)
.RE
.PP
\fB\-t\fR \fIT\fR, \fB\-\-tries\fR \fIT\fR
.RS 4
The number of times to try to execute the task before failing permanently\&. This is the task\-level equivalent of the
\fB\-\-tries\fR
command\-line option\&.
.RE
.PP
\fB\-p\fR \fIP\fR, \fB\-\-priority\fR \fIP\fR
.RS 4
The priority of the task\&. P should be an integer\&. Larger values have higher priority\&. The default is 0\&. Priorities are simply hints and are not strict\(emif a task cannot be matched to an available slot (e\&.g\&. due to resource availability), but a lower\-priority task can, then the task will be deferred and the lower priority task will be executed\&. This option can be set for a job in the DAX by specifying the pegasus::pmc_priority profile\&.
.RE
.PP
\fB\-f\fR \fIVAR=FILE\fR, \fB\-\-pipe\-forward\fR \fIVAR=FILE\fR
.RS 4
Forward I/O to file
\fIFILE\fR
using pipes to communicate with the task\&. The environment variable
\fIVAR\fR
will be set to the value of a file descriptor for a pipe to which the task can write to get data into
\fIFILE\fR\&. For example, if a task specifies: \-f FOO=/tmp/foo then the environment variable FOO for the task will be set to a number (e\&.g\&. 3) that represents the file /tmp/foo\&. In order to specify this argument in a Pegasus DAX you need to set the pegasus::pmc_arguments profile (note that the value of pmc_arguments must contain the "\-f" part of the argument, so a valid value would be: <profile namespace="pegasus" key="pmc_arguments">\-f A=/tmp/a </profile>)\&. (see
\fBI/O FORWARDING\fR)
.RE
.PP
\fB\-F\fR \fISRC=DEST\fR, \fB\-\-file\-forward\fR \fISRC=DEST\fR
.RS 4
Forward I/O to the file
\fIDEST\fR
from the file
\fISRC\fR\&. When the task finishes, the worker will read the data from
\fISRC\fR
and send it to the master where it will be written to the file
\fIDEST\fR\&. After
\fISRC\fR
is read it is deleted\&. In order to specify this argument in a Pegasus DAX you need to set the pegasus::pmc_arguments profile\&. (see
\fBI/O FORWARDING\fR)
.RE
.sp
The format of an \fBEDGE\fR record is:
.sp
.if n \{\
.RS 4
.\}
.nf
"EDGE" parent child
.fi
.if n \{\
.RE
.\}
.sp
Where \fIparent\fR is the ID of the parent task, and \fIchild\fR is the ID of the child task\&. An example \fBEDGE\fR record is:
.sp
.if n \{\
.RS 4
.\}
.nf
EDGE t01 t02
.fi
.if n \{\
.RE
.\}
.sp
A simple diamond\-shaped workflow would look like this:
.sp
.if n \{\
.RS 4
.\}
.nf
# diamond\&.dag
TASK A /bin/echo "I am A"
TASK B /bin/echo "I am B"
TASK C /bin/echo "I am C"
TASK D /bin/echo "I am D"

EDGE A B
EDGE A C
EDGE B D
EDGE C D
.fi
.if n \{\
.RE
.\}
.SH "RESCUE FILES"
.sp
Many different types of errors can occur when running a DAG\&. One or more of the tasks may fail, the MPI job may run out of wall time, \fBpegasus\-mpi\-cluster\fR may segfault (we hope not), the system may crash, etc\&. In order to ensure that the DAG does not need to be restarted from the beginning after an error, \fBpegasus\-mpi\-cluster\fR generates a rescue file for each workflow\&.
.sp
The rescue file is a simple text file that lists all of the tasks in the workflow that have finished successfully\&. This file is updated each time a task finishes, and is flushed periodically so that if the work\- flow fails and the user restarts it, \fBpegasus\-mpi\-cluster\fR can determine which tasks still need to be executed\&. As such, the rescue file is a sort\-of transaction log for the workflow\&.
.sp
The rescue file contains zero or more DONE records\&. The format of these records is:
.sp
.if n \{\
.RS 4
.\}
.nf
"DONE" *taskid*
.fi
.if n \{\
.RE
.\}
.sp
Where \fItaskid\fR is the ID of the task that finished successfully\&.
.sp

RELEASE_NOTES
$Id$
===============================
Release Notes for PEGASUS 4.3.1
===============================
We are happy to announce the release of Pegasus 4.3.1 .  

Pegasus 4.3.1 is a minor release, that has minor enhancements and fixes bugs to Pegasus 4.3.0 release. 
 
Enhancements:
--------------

1) Support for Fixed Output Mapper

   Using this output mapper, users can specify  an externally
   accessible URL in the properties file, pointing to a directory
   where the output files needs to be transferred to. To use this
   mapper, set the following  properties  
 
	pegasus.dir.storage.mapper Fixed
	pegasus.dir.storage.mapper.fixed.url  <url to the storage directory e.g. gsiftp://outputs.isi.edu/shared/outputs>
 

Bugs Fixed:
-------------- 

1) pegasus-analyzer does not detect jobs that are condor_rm'ed if no
postscript is associated with the job 

   By default, each job has a postscript associated that detects an
   empty job stdout and flags it as a failure. However, if a job is
   not asscociated with a postscript and a user/system condor_rm's the
   job, the failure is not detected. This is now fixed, and a
   JOB_ABORTED event is logged in the stampede database, when a job is
   aborted. 
 
2) IRODS support in pegasus-transfer

   The IRODS support in pegasus-transfer was broken. This is now fixed.
 
3) pegasus-kickstart compilation warnings for character encodings

   kickstart maintains a table to escape characters correctly for
   putting them in a XML document. The non-ascii characters in the
   table were latin1 not UTF-8. This caused a warning on newer
   versions of gcc, which could not be disabled across all platforms.
   kickstart now writes out it;s output in UTF-8 encoding and the xml
   escaping was updated accordingly. 
 
4) Fix to URL handling in the planner

   Changed the regex in PegasusURL so that we can pass urls with just
   the hostnames specified. e.g http://isis.isi.edu Note: no trailing
   / . Before the fix, the planner was throwing an exception if a user
   specified an input URL with path names containing only one
   directory. 
 
5) planner had a rogue debug statement

   There was a rogue system.out statement in the planner output that
   led to a statement being logged for each job in the workflow. 
 
6) pegasus-statistics had 2.6 style code

   Pegasus is distributed as part of OSG software stack, and one of
   the supported platforms there is EL5 systems that come with Python
   2.4. pegasus-statistics had some code that was compatible with
   2.5. This is now fixed.

===============================
Release Notes for PEGASUS 4.3
===============================
We are happy to announce the release of Pegasus 4.3.  

Pegasus 4.3 is a major release of Pegasus which contains all the
enhancements and bugfixes in 4.2.2 release 

New features and Improvements in 4.3 include

    - improvements to pegasus lite and optimizations for input file
    staging in non shared filesystem deployments 
    - support for output mappers, allowing users finer grained control
    over where to place the outputs on an output site 
    - support for SSH based submissions on top of Condor BOSCO. 
    - substantial improvements to pegasus-kickstart including abiltiy
    to track peak memory usage for jobs 
    - improvements to pegasus-s3 and pegasus-transfer


NEW FEATURES
--------------

1) Support for bypassing transfer of input files via the staging site 

   In the non shared filesystem deployments (
   pegasus.data.configuration = nonsharedfs|condorio) users, can now
   setup pegasus to transfer the input files directly to the worker
   nodes without going through the staging site.  This can be done by
   setting the following property to true 
  
   pegasus.transfer.bypass.input.staging

   In the nonsharedfs case, if the input files are already present on
   a shared disk accessible from the worker nodes, pegasus lite can
   symlink instead of copying them over to the local directory on the
   worker node.  The cleanup algorithm was updated to ignore files
   that are directly pulled to the worker nodes from the input site
   locations. 
 
2) Support for DAX generation jobs in hierarchal workflows

   Pegasus now has support for having a dax generation job in the
   workflow. This allows users to add long running dax generation jobs
   as a compute job in the workflow, that can be run remotely. These
   dax generation jobs need to be a parent of the associated DAX
   job. Pegasus will ensure that the DAX generated on a remote site is
   brought to the local site for the associated sub workflow
   corresponding to the DAX job to be planned. 

   Earlier, the only way for hierarchal workflows was that the DAX'es
   for the sub workflows had to be pre generated and the paths to the
   dax files was specified in the DAX jobs.  Pegasus did not
   automatically handle separate DAX generations jobs out of the
   box. More details can be found at
   https://jira.isi.edu/browse/PM-667. There is an example checked in
   the share/pegasus/examples/dynamic-hierarchy directory. 
 
3) Support for output mappers
   
   Pegasus now has support for output mappers, that allow users fine
   grained control over how the output files on the output site are
   laid out.  The mappers can be configured by setting the following
   property  

    pegasus.dir.storage.mapper

   The following mappers are supported

   Flat: By default, Pegasus will place the output files in the
   storage directory specified in the site catalog for the output
   site. 

   Hashed: This mapper results in the creation of a deep directory
   structure on the output site, while populating the results. The
   base directory on the remote end is determined from the site
   catalog. Depending on the number of files being staged to the
   remote site a Hashed File Structure is created that ensures that
   only 256 files reside in one directory. To create this directory

src/tools/pegasus-mpi-cluster/pegasus-mpi-cluster.cpp
/* mpi.h must come before stdio.h for Intel MPI */
#include <mpi.h>
#include <stdio.h>
#include <list>
#include <stdlib.h>
#include <unistd.h>
#include <math.h>

#include "version.h"
#include "dag.h"
#include "engine.h"
#include "master.h"
#include "worker.h"
#include "failure.h"
#include "log.h"
#include "mpicomm.h"
#include "protocol.h"
#include "tools.h"

using std::string;
using std::list;
using std::exception;

static char *program = NULL;
static int rank = 0;

void version() {
    if (rank == 0) {
#ifdef PEGASUS_VERSION
        fprintf(stderr, "Version: %s\n", PEGASUS_VERSION);
#endif
        fprintf(stderr, "Compiled: %s %s\n", __DATE__, __TIME__);
#ifdef __VERSION__
        fprintf(stderr, "Compiler: %s\n", __VERSION__);
#endif
        int major, minor;
        MPI_Get_version(&major, &minor);
        fprintf(stderr, "MPI: %d.%d\n", major, minor);
#ifdef MPICH_VERSION
        fprintf(stderr, "MPICH: %s\n", MPICH_VERSION);
#endif
#ifdef MPICH2_VERSION
        fprintf(stderr, "MPICH2: %s\n", MPICH2_VERSION);
#endif
#ifdef MVAPICH2_VERSION
        fprintf(stderr, "MVAPICH2: %s\n", MVAPICH2_VERSION);
#endif
#ifdef OPEN_MPI
        fprintf(stderr, "OpenMPI: %d.%d.%d\n", OMPI_MAJOR_VERSION, OMPI_MINOR_VERSION, OMPI_RELEASE_VERSION);
#endif
    }
}

void usage() {
    if (rank == 0) {
        fprintf(stderr,
            "Usage: %s [options] DAGFILE\n"
            "\n"
            "Options:\n"
            "   -h|--help            Print this message\n"
            "   -V|--version         Print version information\n"
            "   -v|--verbose         Increase logging level\n"
            "   -q|--quiet           Decrease logging level\n"
            "   -o|--stdout PATH     Path to stdout file for tasks\n"
            "   -e|--stderr PATH     Path to stderr file for tasks\n"
            "   -s|--skip-rescue     Ignore existing rescue file (still creates one)\n"
            "   -m|--max-failures N  Stop submitting tasks after N tasks have failed\n"
            "   -t|--tries N         Try tasks N times before marking them failed\n"
            "   -n|--nolock          Do not try to lock DAGFILE\n"
            "   -r|--rescue PATH     Path to rescue log [default: DAGFILE.rescue]\n"
            "   --host-script PATH   Path to script that will be launched on each host\n"
            "   --host-memory N      Amount of memory per host in MB\n"
            "   --host-cpus N        Number of CPUs per host\n"
            "   --strict-limits      Enforce strict task resource limits\n"
            "   --max-wall-time T    Maximum wall time of the job in minutes\n"
            "   --per-task-stdio     Write each task's stdout/stderr to a different file\n"
            "   --jobstate-log       Generate jobstate.log\n"
            "   --monitord-hack      Generate a .dagman.out file to trick monitord\n"
            "   --no-resource-log    Do not generate a log of resource usage\n"
            "   --no-sleep-on-recv   Do not sleep on message receive\n"
            "   --maxfds             Maximum cached file descriptors\n",
            program
        );
    }
}

void argerror(const string message) {
    if (rank == 0) {
        fprintf(stderr, "%s\n", message.c_str());
    }
}

int mpidag(int argc, char *argv[], MPICommunicator &comm) {
    rank = comm.rank();
    int numprocs = comm.size();

    list<char *> flags;
    for (int i=1; i<argc; i++) {
        flags.push_back(argv[i]);
    }

    string outfile = "stdout";
    string errfile = "stderr";
    list<string> args;
    int loglevel = LOG_INFO;
    bool skiprescue = false;
    int max_failures = 0;
    int tries = 1;
    bool lock = true;
    string rescuefile = "";
    string host_script = "";
    unsigned host_memory = 0;
    unsigned host_cpus = 0;
    bool strict_limits = false;
    double max_wall_time = 0.0;
    bool per_task_stdio = false;
    bool jobstate_log = false;
    bool monitord_hack = false;
    bool log_resources = true;
    bool sleep_on_recv = true;
    int maxfds = 0;

    // Environment variable defaults
    char *env_host_script = getenv("PMC_HOST_SCRIPT");
    if (env_host_script != NULL) {
        host_script = env_host_script;
    }

    char *env_host_memory = getenv("PMC_HOST_MEMORY");
    if (env_host_memory != NULL) {
        if (sscanf(env_host_memory, "%u", &host_memory) != 1) {
            argerror("Invalid value for PMC_HOST_MEMORY");
            return 1;
        }
    }

    char *env_host_cpus = getenv("PMC_HOST_CPUS");
    if (env_host_cpus != NULL) {
        if (sscanf(env_host_cpus, "%u", &host_cpus) != 1) {
            argerror("Invalid value for PMC_HOST_CPUS");
            return 1;
        }
    }

    char *env_max_wall_time = getenv("PMC_MAX_WALL_TIME");
    if (env_max_wall_time != NULL) {
        if (sscanf(env_max_wall_time, "%lf", &max_wall_time) != 1) {
            argerror("Invalid value for PMC_MAX_WALL_TIME");
            return 1;
        }
    }

    while (flags.size() > 0) {
        string flag = flags.front();
        if (flag == "-h" || flag == "--help") {
            // This should no longer be reachable because
            // of the code in main()
            usage();
            return 0;
        } else if (flag == "-V" || flag == "--version") {
            version();
            return 0;
        } else if (flag == "-o" || flag == "--stdout") {
            flags.pop_front();
            if (flags.size() == 0) {
                argerror("-o/--stdout requires PATH");
                return 1;
            }
            outfile = flags.front();
        } else if (flag == "-e" || flag == "--stderr") {
            flags.pop_front();
            if (flags.size() == 0) {
                argerror("-e/--stderr requires PATH");
                return 1;
            }
            errfile = flags.front();
        } else if (flag == "-q" || flag == "--quiet") {
            loglevel -= 1;
        } else if (flag == "-v" || flag == "--verbose") {
            loglevel += 1;
        } else if (flag == "-s" || flag == "--skip-rescue") {
            skiprescue = true;
        } else if (flag == "-m" || flag == "--max-failures") {
            flags.pop_front();
            if (flags.size() == 0) {
                argerror("-m/--max-failures requires N");
                return 1;
            }
            string N = flags.front();
            if (!sscanf(N.c_str(), "%d", &max_failures)) {
                argerror("N for -m/--max-failures is invalid");
                return 1;
            }
            if (max_failures < 0) {
                argerror("N for -m/--max-failures must be >= 0");
                return 1;
            }
        } else if (flag == "-t" || flag == "--tries") {
            flags.pop_front();
            if (flags.size() == 0) {
                argerror("-t/--tries requires N");
                return 1;
            }
            string N = flags.front();
            if (!sscanf(N.c_str(), "%d", &tries)) {
                argerror("N for -t/--tries is invalid");
                return 1;
            }
            if (tries < 1) {
                argerror("N for -t/--tries must be >= 1");
                return 1;
            }
        } else if (flag == "-n" || flag == "--nolock") {
            lock = false;
        } else if (flag == "-r" || flag == "--rescue") {
            flags.pop_front();
            if (flags.size() == 0) {
                argerror("-r/--rescue requires PATH");
                return 1;
            }
            rescuefile = flags.front();
        } else if (flag == "--host-script") {
            flags.pop_front();
            if (flags.size() == 0) {
                argerror("--host-script requires PATH");
                return 1;
            }
            host_script = flags.front();
        } else if (flag == "--host-memory") {
            flags.pop_front();
            if (flags.size() == 0) {
                argerror("--host-memory requires N");
                return 1;
            }
            string host_memory_string = flags.front();
            if (sscanf(host_memory_string.c_str(), "%u", &host_memory) != 1) {
                argerror("Invalid value for --host-memory");
                return 1;
            }
            if (host_memory < 1) {
                argerror("--host-memory must be an integer >= 1");
                return 1;
            }
        } else if (flag == "--host-cpus") {
            flags.pop_front();
            if (flags.size() == 0) {
                argerror("--host-cpus requires N");
                return 1;
            }
            string host_cpus_string = flags.front();
            if (sscanf(host_cpus_string.c_str(), "%u", &host_cpus) != 1) {
                argerror("Invalid value for --host-cpus");
                return 1;
            }
            if (host_cpus < 1) {
                argerror("--host-cpus must be an integer >= 1");
                return 1;
            }
        } else if (flag == "--strict-limits") {
            strict_limits = true;
        } else if (flag == "--max-wall-time") {
            flags.pop_front();
            if (flags.size() == 0) {
                argerror("--max-wall-time requires T");
                return 1;
            }
            string max_wall_time_string = flags.front();
            if (sscanf(max_wall_time_string.c_str(), "%lf", &max_wall_time) != 1) {
                argerror("Invalid value for --max-wall-time");
                return 1;
            }
        } else if (flag == "--per-task-stdio") {
            per_task_stdio = true;
        } else if (flag == "--jobstate-log") {
            jobstate_log = true;
        } else if (flag == "--monitord-hack") {
            monitord_hack = true;
            per_task_stdio = true;
        } else if (flag == "--no-resource-log") {
            log_resources = false;
        } else if (flag == "--no-sleep-on-recv") {
            sleep_on_recv = false;
        } else if (flag == "--maxfds") {
            flags.pop_front();
            if (flags.size() == 0) {
                argerror("--maxfds requires N");
                return 1;
            }
            string maxfds_string = flags.front();
            if (sscanf(maxfds_string.c_str(), "%d", &maxfds) != 1) {
                argerror("Invalid value for --maxfds");
                return 1;
            }
            if (maxfds < 1) {
                argerror("--maxfds must be at least 1");
                return 1;
            }
        } else if (flag[0] == '-') {
            string message = "Unrecognized argument: ";
            message += flag;
            argerror(message);
            return 1;
        } else {
            args.push_back(flag);
        }
        flags.pop_front();
    }

    if (args.size() == 0) {
        usage();
        return 1;
    }

    if (args.size() > 1) {
        argerror("Invalid argument\n");
        return 1;
    }

    string dagfile = args.front();

    log_set_level(loglevel);

    if (numprocs < 2) {
        fprintf(stderr, "At least one worker process is required\n");
        return 1;
    }

    comm.sleep_on_recv = sleep_on_recv;

    // Everything is pretty deterministic up until the processes reach
    // this point. Once we get here the different processes can diverge 
    // in their behavior for many reasons (file systems issues, bad nodes,
    // etc.), so be careful how failures are handled after this point
    // and make sure MPI_Abort is called when something bad happens.

    if (rank == 0) {
        version();

        // If no rescue file specified, use default
        if (rescuefile == "") {
            rescuefile = dagfile + ".rescue";
        }

        string oldrescue = rescuefile;
        string newrescue = rescuefile;

        if (skiprescue) {
            // User does not want to read old rescue file
            oldrescue = "";
        }

        log_debug("Using old rescue file: %s", oldrescue.c_str());
        log_debug("Using new rescue file: %s", newrescue.c_str());

        string resource_log;
        if (log_resources) {
            resource_log = dagfile + ".resource";
        }

        bool has_host_script = ("" != host_script);

        DAG dag(dagfile, oldrescue, lock, tries);
        Engine engine(dag, newrescue, max_failures);
        Master master(&comm, program, engine, dag, dagfile, outfile, errfile,
                has_host_script, max_wall_time, resource_log, per_task_stdio,
                maxfds);

        string jobstate_path = dirname(dagfile) + "/jobstate.log";
        JobstateLog jslog(jobstate_path);
        if (jobstate_log) {
            master.add_listener(&jslog);
        }

        DAGManLog dagmanlog(dagfile + ".dagman.out", dagfile);
        if (monitord_hack) {
            master.add_listener(&dagmanlog);
        }

        return master.run();
    } else {

        Worker worker(&comm, dagfile, host_script, host_memory, host_cpus, 
                strict_limits, per_task_stdio);

        return worker.run();
    }
}

void out_of_memory() {
    myfailure("Unable to allocate memory");
}

int main(int argc, char *argv[]) {
    program = argv[0];

    // Shortcut to make 'help' work when running without an MPI context
    if (argc < 2) {
        usage();
        return 1;
    }
    for (int i=1; i<argc; i++) {
        string flag = argv[i];
        if (flag == "-h" || flag == "--help") {
            usage();
            return 0;
        }
    }

    MPICommunicator comm(&argc, &argv);
    try {
        std::set_new_handler(out_of_memory);
        int rc = mpidag(argc, argv, comm);
        return rc;
    } catch (exception &error) {
        // If we catch an execption here, then one of the
        // processes has hit an unsolvable problem and we
        // need to abort the entire workflow.
        fprintf(stderr, "ABORT: %s\n", error.what());
        comm.abort(1);
    }
}

